File: app/src/main/java/de/jerleo/database/TableTariff.java
package de.jerleo.database;

import android.content.ContentValues;
import android.database.Cursor;

import java.util.ArrayList;
import java.util.List;

import de.jerleo.android.DateHelper;
import de.jerleo.database.DatabaseColumn.DataType;
import de.jerleo.model.Tariff;

import static android.provider.BaseColumns._ID;
import static de.jerleo.database.Constants.DATE;
import static de.jerleo.database.Constants.FEE;
import static de.jerleo.database.Constants.METER;
import static de.jerleo.database.Constants.PAYMENT;
import static de.jerleo.database.Constants.PRICE;
import static de.jerleo.database.Constants.TARIFF;

public class TableTariff extends DatabaseTable {

    public TableTariff(Database database) {

        super(database, TARIFF, true);

        addColumn(new DatabaseColumn(METER, DataType.INTEGER));
        addColumn(new DatabaseColumn(DATE, DataType.TEXT));
        addColumn(new DatabaseColumn(FEE, DataType.REAL));
        addColumn(new DatabaseColumn(PRICE, DataType.REAL));
        addColumn(new DatabaseColumn(PAYMENT, DataType.REAL));

        final DatabaseColumn meter = getColumn(METER);
        meter.setForeignKeyConstraint(true);
        meter.setForeignTable(METER);
        meter.setForeignColumn(_ID);
    }

    @Override
    public ContentValues getValues(Object object, boolean forUpdate) {

        final Tariff tariff = (Tariff) object;

        final ContentValues values = new ContentValues();
        if (forUpdate)
            values.put(_ID, tariff.getId());
        values.put(METER, tariff.getMeter().getId());
        values.put(DATE, DateHelper.getDatabaseDate(tariff.getValidFrom()));
        values.put(FEE, tariff.getFee());
        values.put(PRICE, tariff.getPrice());
        values.put(PAYMENT, tariff.getPayment());

        return values;
    }

    @Override
    public String getWhereClause(Object object) {

        final Tariff tariff = (Tariff) object;
        return _ID + " = " + tariff.getId();
    }

    @Override
    public List<Tariff> read(String condition) {

        final ArrayList<Tariff> tariffs = new ArrayList<>();
        final String[] columns = {_ID, DATE, FEE, PRICE, PAYMENT};
        final String where = METER + " = " + condition;
        final Cursor cursor = database.query(TARIFF, columns, where, null,
                null, null, null);
        while (cursor.moveToNext()) {
            final Tariff tariff = new Tariff();
            int col = 0;
            tariff.setId(cursor.getLong(col++));
            tariff.setValidFrom(DateHelper.getDatabaseDate(cursor.getString(col++)));
            tariff.setFee(cursor.getFloat(col++));
            tariff.setPrice(cursor.getFloat(col++));
            tariff.setPayment(cursor.getFloat(col));
            tariffs.add(tariff);
        }
        return tariffs;
    }

}

File: app/src/main/java/de/jerleo/database/table/TableTariff.kt
package de.jerleo.database.table

import android.content.ContentValues
import android.provider.BaseColumns
import de.jerleo.android.DateHelper
import de.jerleo.database.Column
import de.jerleo.database.Constants
import de.jerleo.database.Database
import de.jerleo.database.Table
import de.jerleo.model.Tariff

class TableTariff(database: Database) :
    Table(database, Constants.TARIFF, true) {

    init {
        add(Column(Constants.METER, Column.Type.INTEGER).apply {
            isForeignKeyConstraint = true
            foreignTable = Constants.METER
            foreignColumn = BaseColumns._ID
        })
        add(Column(Constants.DATE, Column.Type.TEXT))
        add(Column(Constants.FEE, Column.Type.REAL))
        add(Column(Constants.PRICE, Column.Type.REAL))
        add(Column(Constants.PAYMENT, Column.Type.REAL))
    }

    override fun values(obj: Any, forUpdate: Boolean): ContentValues {
        val tariff = obj as Tariff
        val values = ContentValues()
        if (forUpdate) values.put(BaseColumns._ID, tariff.id)
        values.put(Constants.METER, tariff.meter.id)
        values.put(Constants.DATE, tariff.dateFrom.let { DateHelper.format(it) })
        values.put(Constants.FEE, tariff.fee)
        values.put(Constants.PRICE, tariff.price)
        values.put(Constants.PAYMENT, tariff.payment)
        return values
    }

    override fun whereClause(obj: Any): String = BaseColumns._ID + " = " + (obj as Tariff).id

    override fun read(condition: String?): ArrayList<Tariff> {
        val tariffs = ArrayList<Tariff>()
        val columns = arrayOf(
            BaseColumns._ID,
            Constants.DATE,
            Constants.FEE,
            Constants.PRICE,
            Constants.PAYMENT
        )
        val where = Constants.METER + " = " + condition
        val cursor = database.query(
            Constants.TARIFF, columns, where, null,
            null, null, null
        )
        while (cursor.moveToNext()) {
            val tariff = Tariff()
            var col = 0
            tariff.id = cursor.getLong(col++)
            tariff.dateFrom = DateHelper.parse(cursor.getString(col++))
            tariff.fee = cursor.getDouble(col++)
            tariff.price = cursor.getDouble(col++)
            tariff.payment = cursor.getDouble(col)
            tariffs.add(tariff)
        }
        return tariffs
    }
}